File: backend/app/services/search_orchestrator.py
from typing import List, Dict, Any
from backend.app.services.query_translator import QueryTranslator
from backend.app.services.api_integrator import ApiIntegrator

class SearchOrchestrator:
    def __init__(self, db_manager):
        self.db_manager = db_manager
        self.translator = QueryTranslator()
        self.integrator = ApiIntegrator()

    async def search_and_persist(self, project_id: int, query_blocks: List[Dict], limit: int = 100) -> int:
        # 1. Translate Query
        openalex_filter = self.translator.to_openalex(query_blocks)
        crossref_params = self.translator.to_crossref(query_blocks)
        
        # 2. Fetch results
        oa_raw = await self.integrator.fetch_openalex(openalex_filter)
        cr_raw = await self.integrator.fetch_crossref(crossref_params)
        
        # 3. Normalize
        normalized_results = []
        for item in oa_raw[:limit]:
            norm = self.integrator.normalize_openalex(item)
            norm["base_origem"] = "OpenAlex"
            normalized_results.append(norm)
            
        for item in cr_raw[:limit]:
            norm = self.integrator.normalize_crossref(item)
            norm["base_origem"] = "Crossref"
            normalized_results.append(norm)
            
        # 4. Deduplicate
        deduplicated = self._deduplicate(normalized_results)
        
        # 5. Persist
        count = 0
        for article in deduplicated:
            # Map CSL-JSON to DB fields
            article_db = {
                "projeto_id": project_id,
                "doi": article.get("DOI"),
                "titulo": article.get("title"),
                "autores": ", ".join([f"{a.get('given', '')} {a.get('family', '')}".strip() for a in article.get("author", [])]),
                "ano": article.get("issued", {}).get("date-parts", [[None]])[0][0],
                "query_origem": str(query_blocks),
                "base_origem": article["base_origem"], # This is a list from _deduplicate
                "csl_json": article
            }
            self.db_manager.save_article(article_db)
            count += 1
            
        return count

    def _deduplicate(self, results: List[Dict]) -> List[Dict]:
        seen_doi = {} # DOI -> index in deduplicated
        seen_title = {} # Title.lower() -> index in deduplicated
        deduplicated = []
        
        for item in results:
            doi = item.get("DOI")
            title = item.get("title", "").lower().strip()
            
            existing_idx = None
            if doi and doi in seen_doi:
                existing_idx = seen_doi[doi]
            elif title and title in seen_title:
                existing_idx = seen_title[title]
            
            if existing_idx is not None:
                # Merge base_origem
                current_bases = deduplicated[existing_idx]["base_origem"]
                if isinstance(current_bases, str):
                    current_bases = [current_bases]
                
                new_base = item["base_origem"]
                if new_base not in current_bases:
                    current_bases.append(new_base)
                
                deduplicated[existing_idx]["base_origem"] = current_bases
            else:
                idx = len(deduplicated)
                item["base_origem"] = [item["base_origem"]]
                deduplicated.append(item)
                if doi:
                    seen_doi[doi] = idx
                if title:
                    seen_title[title] = idx
                    
        return deduplicated

File: frontend/electron/services/SearchOrchestrator.ts
import { DatabaseManager } from '../database/DatabaseManager';
import { QueryTranslator } from './QueryTranslator';
import { ApiIntegrator } from './ApiIntegrator';
import { QueryBlock, NormalizedArticle } from './types';
import { Article } from '../types';

export class SearchOrchestrator {
  constructor(
    private db: DatabaseManager,
    private translator: QueryTranslator,
    private api: ApiIntegrator
  ) {}

  async executeSearch(
    projectId: number,
    queryBlocks: QueryBlock[],
    limit: number = 100
  ): Promise<{ savedCount: number; articles: Article[] }> {
    const openalexFilter = this.translator.toOpenAlex(queryBlocks);
    const crossrefParams = this.translator.toCrossref(queryBlocks);
    
    // Fetch results
    const [oaResults, crResults] = await Promise.all([
      this.api.searchOpenAlex(openalexFilter),
      this.api.searchCrossref(crossrefParams)
    ]);
    
    const combinedResults = [...oaResults.slice(0, limit), ...crResults.slice(0, limit)];
    
    const deduplicated = this.deduplicate(combinedResults);
    
    let savedCount = 0;
    for (const article of deduplicated) {
      this.db.saveArticle(projectId, {
        doi: article.doi,
        title: article.title,
        authors: article.authors,
        year: article.year,
        source_query: JSON.stringify(queryBlocks),
        source_databases: JSON.stringify(article.source_databases),
        csl_json: JSON.stringify(article.csl_json)
      });
      savedCount++;
    }
    
    const projectArticles = this.db.getArticlesByProject(projectId);
    return { savedCount, articles: projectArticles };
  }

  private deduplicate(results: NormalizedArticle[]): NormalizedArticle[] {
    const seenDoi = new Map<string, number>();
    const seenTitle = new Map<string, number>();
    const deduplicated: NormalizedArticle[] = [];
    
    for (const item of results) {
      const doi = item.doi;
      const title = (item.title || "").toLowerCase().trim();
      
      let existingIdx: number | undefined;
      
      if (doi && seenDoi.has(doi)) {
        existingIdx = seenDoi.get(doi);
      } else if (title && seenTitle.has(title)) {
        existingIdx = seenTitle.get(title);
      }
      
      if (existingIdx !== undefined) {
        const existing = deduplicated[existingIdx];
        const newSource = item.source_databases[0];
        if (!existing.source_databases.includes(newSource)) {
          existing.source_databases.push(newSource);
        }
      } else {
        const idx = deduplicated.length;
        deduplicated.push(item);
        if (doi) seenDoi.set(doi, idx);
        if (title) seenTitle.set(title, idx);
      }
    }
    
    return deduplicated;
  }
}
